server/src/main/java/org/elasticsearch/index/codec/vectors/DefaultIVFVectorsWriter.java

```diff
@@ -0,0 +1,686 @@
+        }
+    }
+
+    @Override
+    CentroidAssignmentScorer createCentroidScorer(
+        IndexInput centroidsInput,
+        int numCentroids,
+        FieldInfo fieldInfo,
+        float[] globalCentroid
+    ) {
+        return new OffHeapCentroidAssignmentScorer(centroidsInput, numCentroids, fieldInfo);
+    }
+
+    static void writeCentroids(float[][] centroids, FieldInfo fieldInfo, float[] globalCentroid, IndexOutput centroidOutput)
+        throws IOException {
+        final OptimizedScalarQuantizer osq = new OptimizedScalarQuantizer(fieldInfo.getVectorSimilarityFunction());
+        byte[] quantizedScratch = new byte[fieldInfo.getVectorDimension()];
+        float[] centroidScratch = new float[fieldInfo.getVectorDimension()];
+        // TODO do we want to store these distances as well for future use?
+        float[] distances = new float[centroids.length];
+        for (int i = 0; i < centroids.length; i++) {
+            distances[i] = VectorUtil.squareDistance(centroids[i], globalCentroid);
+        }
+        // sort the centroids by distance to globalCentroid, nearest (smallest distance), to furthest
+        // (largest)
+        for (int i = 0; i < centroids.length; i++) {
+            for (int j = i + 1; j < centroids.length; j++) {
+                if (distances[i] > distances[j]) {
+                    float[] tmp = centroids[i];
+                    centroids[i] = centroids[j];
+                    centroids[j] = tmp;
+                    float tmpDistance = distances[i];
+                    distances[i] = distances[j];
+                    distances[j] = tmpDistance;
+                }
+            }
+        }
+        for (float[] centroid : centroids) {
+            System.arraycopy(centroid, 0, centroidScratch, 0, centroid.length);
+            OptimizedScalarQuantizer.QuantizationResult result = osq.scalarQuantize(
+                centroidScratch,
+                quantizedScratch,
+                (byte) 4,
+                globalCentroid
+            );
+            writeQuantizedValue(centroidOutput, quantizedScratch, result);
+        }
+        final ByteBuffer buffer = ByteBuffer.allocate(fieldInfo.getVectorDimension() * Float.BYTES).order(ByteOrder.LITTLE_ENDIAN);
+        for (float[] centroid : centroids) {
+            buffer.asFloatBuffer().put(centroid);
+            centroidOutput.writeBytes(buffer.array(), buffer.array().length);
+        }
+    }
+
+    record SegmentCentroid(int segment, int centroid, int centroidSize) {}
+
+    @Override
+    protected int calculateAndWriteCentroids(
+        FieldInfo fieldInfo,
+        FloatVectorValues floatVectorValues,
+        IndexOutput temporaryCentroidOutput,
+        MergeState mergeState,
+        float[] globalCentroid
+    ) throws IOException {
+        if (floatVectorValues.size() == 0) {
+            return 0;
+        }
+        int desiredClusters = ((floatVectorValues.size() - 1) / vectorPerCluster) + 1;
```

Comment:
Is there a reason why calculation of `desiredClusters` is different from when we are doing `flush`?